python/lsst/pipe/tasks/dcrAssembleCoadd.py

```diff
@@ -0,0 +1,980 @@
+
+        Parameters
+        ----------
+        visitInfo : lsst.afw.image.VisitInfo
+            Metadata for the exposure.
+        wcs : lsst.afw.geom.skyWcs.skyWcs.SkyWcs
+            Coordinate system definition (wcs) for the exposure.
+
+        Returns
+        -------
+        lsst.afw.geom.Extent2I
+            The 2D shift due to DCR, in pixels.
+        """
+        rotation = self.calculateRotationAngle(visitInfo, wcs)
+        dcrShift = []
+        lambdaEff = self.filterInfo.getFilterProperty().getLambdaEff()
+        for wl0, wl1 in self.wavelengthGenerator():
+            # Note that diffRefractAmp can be negative, since it's relative to the midpoint of the full band
+            diffRefractAmp0 = differentialRefraction(wl0, lambdaEff,
+                                                     elevation=visitInfo.getBoresightAzAlt().getLatitude(),
+                                                     observatory=visitInfo.getObservatory(),
+                                                     weather=visitInfo.getWeather())
+            diffRefractAmp1 = differentialRefraction(wl1, lambdaEff,
+                                                     elevation=visitInfo.getBoresightAzAlt().getLatitude(),
+                                                     observatory=visitInfo.getObservatory(),
+                                                     weather=visitInfo.getWeather())
+            diffRefractAmp = (diffRefractAmp0 + diffRefractAmp1)/2.
+            diffRefractPix = diffRefractAmp.asArcseconds()/wcs.getPixelScale().asArcseconds()
+            dcrShift.append(afwGeom.Extent2D(diffRefractPix*np.cos(rotation.asRadians()),
+                                             diffRefractPix*np.sin(rotation.asRadians())))
+        return dcrShift
+
+    def buildMatchedTemplate(self, dcrModels, visitInfo, bbox, wcs, mask=None):
+        """Create a DCR-matched template for an exposure.
+
+        Parameters
+        ----------
+        dcrModels : list of lsst.afw.image.maskedImageF
+            A list of masked images, each containing the model for one subfilter
+        visitInfo : lsst.afw.image.VisitInfo
+            Metadata for the exposure.
+        bbox : lsst.afw.geom.box.Box2I
+            Sub-region of the coadd
+        wcs : lsst.afw.geom.skyWcs.skyWcs.SkyWcs
+            Coordinate system definition (wcs) for the exposure.
+        mask : lsst.afw.image.mask, optional
+            reference mask to use for the template image.
+
+        Returns
+        -------
+        lsst.afw.image.maskedImageF
+            The DCR-matched template
+        """
+        dcrShift = self.dcrShiftCalculate(visitInfo, wcs)
+        templateImage = afwImage.MaskedImageF(bbox)
+        for dcr, model in zip(dcrShift, dcrModels):
+            templateImage += self.convolveDcrModelPlane(model, dcr, bbox=bbox)
+        if mask is not None:
+            templateImage.setMask(mask[bbox, afwImage.PARENT])
```

Comment:
Does this set the pixels or the pointer to the pixels? I.e., do you have to worry about things changing underneath you?
I'm thinking it's the latter, so yes; that may not be what you want.

Reply:
According to the comment for `setMask` in `MaskedImage.h` "This copies pixel values, not pointers", so luckily it looks like it does the right thing for me.